STORM-1086. Make FailedMsgRetryManager configurable when setting up KafkaSpout

kleinsch opened the pull request:
We'd like to be able to drop messages when they've been replayed N times. Seems easy to do if we could extend ExponentialBackoffMsgRetryManager, but there's no way to inject a custom FailedMsgRetryManager without changing the Kafka codebase. This PR allows users to specify the FailedMsgRetryManager as part of SpoutConfig.

ptgoetz commented on external/storm-kafka/src/jvm/storm/kafka/FailedMsgRetryManagerConfig.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.io.Serializable;
+
+public interface FailedMsgRetryManagerConfig extends Serializable {
```
Since this a factory pattern, I would rename this `FailedMsgRetryManagerFactory`, and the impl `ExponentialBackoffMsgRetryManagerFactory`.

ptgoetz commented on external/storm-kafka/src/jvm/storm/kafka/SpoutConfig.java:
```diff
@@ -39,6 +39,8 @@
     public double retryDelayMultiplier = 1.0;
     public long retryDelayMaxMs = 60 * 1000;
```
Since the above values only apply to the ExponentialBackoff implementation, it would be nice if we could factor them out of this class. They could probably just be moved to `ExponentialBackoffMsgRetryManagerFactory`, and in that case we could probably eliminate the factory pattern altogether.

jerrypeng commented:
non trivial fail in the unit tests that need to be investigated

kleinsch commented:
I wasn't sure if I could make a non-backwards compatible change to SpoutConfig, but seems like a good idea to move the implementation-specific code out of there.

Renamed to FailedMsgRetryManagerFactory and moved params for the exponential backoff manager into the ExponentialBackoffMsgRetryManagerFactory.

This needs to stay a factory bc the SpoutConfig has to be serialized to be sent out to the nodes, then construct the FailedMsgRetryManager per-node. I didn't want to require that the FailedMsgRetryManager be serializable, so said the Factory would be serializable but the RetryManager doesn't need to be.

kleinsch commented:
I've run the tests locally using `mvn install` and they pass using JDK7 and JDK8. Any tips for why the Travis build would pass for Java 8 but not for Java 7?

kleinsch commented:
OK, all feedback addressed and checks passing. Should be ready to merge!

kleinsch commented:
Ping @harshach @ptgoetz @jerrypeng

harshach commented on external/storm-kafka/src/jvm/storm/kafka/FailedMsgRetryManagerFactory.java:
```diff
@@ -0,0 +1,24 @@
+import java.io.Serializable;
+
+public interface FailedMsgRetryManagerFactory extends Serializable {
+    FailedMsgRetryManager construct();
```
can you rename this to build instead of construct.

harshach commented:
@kleinsch re-thinking a bit not sure if we need this Factory construct given that this is single class will only be created once in when the spout initializes. Why not try the following
1. add configure method to FailedMsgRetryManager
public interface FailedMsgRetryManager {
     public void configure(Map<String,Object> config)  or public void configure(SpoutConfig config);
}

2. accept fully qualified class name as pluggable instance ( example look for authorizer config in storm-core)
3. by default create instance of ExponentialBackoffMsgRetryManager  . You can use https://github.com/apache/storm/blob/master/storm-core/src/jvm/backtype/storm/utils/Utils.java#L96
4. call configure method with spoutconfig so that it gets all the available options

Crim commented:
Any chance we can revive this guy from the dead?  Happy to help carry this one across the finish line anyway that I can :)

abhishekagarwal87 commented:
I think its been taken care of by 
https://github.com/apache/storm/pull/1331. can you check?

kleinsch commented:
@abhishekagarwal87 Yes, #1331 looks good to me. If that gets merged, we can discard this PR.
